### PR TITLE
Return default ReductionParams when input has zero elements

### LINKED ISSUE #264: Floating-point exception scheduling reduction of zero-element tensor
Currently, scheduling a reduction with a tensor having zero size results in an unhandled SIGFPE due to 0/0 at https://github.com/NVIDIA/Fuser/blob/main/csrc/scheduler/reduction.cpp#L558 (`n_elems=target_blocks=0`).

Here is a minimal reproduction:
```c++
// Test that 0-dimensional tensors do not break reduction scheduler
TEST_F(NVFuserTest, FusionReduceZeroElementTensor_CUDA) {
  auto fusion = std::make_unique<Fusion>();
  FusionGuard fg(fusion.get());

  std::vector<int64_t> input_shape{3, 4, 0, 5};
  auto tv0 = makeSymbolicTensor(4);
  fusion->addInput(tv0);
  auto tv1 = sum(tv0, {1});
  fusion->addOutput(tv1);

  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
  at::Tensor at_x = at::randn(input_shape, options);
  FusionExecutorCache executor_cache(std::move(fusion));
  auto outputs = executor_cache.runFusionWithInputs({at_x});
  auto t2 = at_x.sum({2});

  auto reduction_params = getReductionHeuristics(fusion.get(), {at_x});
  TORCH_CHECK(reduction_params, "Reduction schedule was not generated!");
  scheduleReduction(fusion.get(), *reduction_params);

  testValidate(
      executor_cache.fusion(), outputs, {at_x}, {t2}, __LINE__, __FILE__);
}
```
Note that the reduction does not need to be along the 0 dimension to trigger the crash. We have a test for zero-dim tensors in the presence of reductions, but it is not actually reducing the zero-dim tensor: https://github.com/NVIDIA/Fuser/blob/13a90b6d18273acc6977269d90cfc28cb5e1c8b0/test/test_gpu2.cpp#L5984.

I tried guarding a few of the `ceilDiv`s in `outerReductionHeuristic` which can avoid the SIGFPE but lead to a kernel segfault, so I'm posting this issue for a little discussion before I try a fix. We may just need to check numel at the beginning of that function and short-circuit there if it's zero.

### PR DESCRIPTION
**Pull request by jacobhinkle**
Fixes #264

This just adds a short-circuit to `{inner,outer}ReductionHeuristic` that determines whether the input tensor to the reduction has any size-0 dimensions. If so, we return the default `ReductionParams` early so that we can safely assume numel>0 for the remainder of the heuristic.

### CODE REVIEW

**Comment by naoyam** [comment]
I think that if size-0 dimensions appear as non-reduction domains, the no-op scheduler should be used, so the only case that could fall in this condition is the non-reduction domains are all non-zero and there's a size-0 reduction dimension. Or, `total_reduction_numel` is zero, while `total_iteration_numel` is non-zero. Am I correct?

**Comment by jacobhinkle** [comment]
!build

**Comment by jacobhinkle** [comment]
> I think that if size-0 dimensions appear as non-reduction domains, the no-op scheduler should be used, so the only case that could fall in this condition is the non-reduction domains are all non-zero and there's a size-0 reduction dimension. Or, `total_reduction_numel` is zero, while `total_iteration_numel` is non-zero. Am I correct?

You're right, the NoOp scheduler will pick this up if the only reductions have zero concrete elements. Reduction scheduler could still run if there are multiple reductions, or as you say, when all the non-reduction extents are non-zero. I am testing both cases here, and I checked that we actually return the starting value when we have all non-zero concrete dimensions. This is how torch's `sum` works, but note that the various reductions in torch each work a bit different from one another:
```
[nav] In [1]: import torch

[ins] In [2]: x = torch.randn(3, 0, 2)

[ins] In [3]: x.sum(1)
Out[3]: 
tensor([[0., 0.],
        [0., 0.],
        [0., 0.]])

[ins] In [4]: x.prod(1)
Out[4]: 
tensor([[1., 1.],
        [1., 1.],
        [1., 1.]])

[ins] In [5]: x.amax(1)
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Cell In[5], line 1
----> 1 x.amax(1)

IndexError: amax(): Expected reduction dim 1 to have non-zero size.

[ins] In [6]: x.sum([0,1])
Out[6]: tensor([0., 0.])

[ins] In [7]: x.prod([0,1])
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[7], line 1
----> 1 x.prod([0,1])

TypeError: prod() received an invalid combination of arguments - got (list), but expected one of:
 * (*, torch.dtype dtype)
 * (int dim, bool keepdim, *, torch.dtype dtype)
 * (name dim, bool keepdim, *, torch.dtype dtype)


[ins] In [8]: x.amax([0, 2])
Out[8]: tensor([])
```
Some accept lists of dims, others (prod) accept only single dimensions. Some don't require numel>0 in those dimensions (sum, prod) and others do (amax). Anyway, I believe NVFuser correctly returns the starting value if the number of reduced elements is zero, and I'm somewhat confident now that the reduction scheduler won't balk at any of the scenarios we've laid out so far.

File: csrc/scheduler/reduction.cpp [hunk 1]
```diff
@@ -77,6 +77,25 @@ std::shared_ptr<ReductionParams> innerReductionHeuristic(
 
   const int64_t n_elems = total_reduction_numel * total_iteration_numel;
 
+  if (n_elems == 0) {
```
Comment (naoyam) [inline csrc/scheduler/reduction.cpp:80]: Instead of having this condition in both `innerReductionHeuristic` and `outerReductionHeuristic`, can we have just one in `reductionHeuristic`?

File: csrc/scheduler/pointwise.cpp [hunk 2]
```diff
@@ -434,11 +434,11 @@ void schedulePointwise(Fusion* fusion, const PointwiseParams& params) {
...
 
-  // maybe has_reduction for scheduling should be done on a per output tensor
-  // basis.
-  TORCH_INTERNAL_ASSERT(
-      ir_utils::getReductionOps(fusion).empty(),
-      "This scheduler only handles pointwise ops.");
+  // Replace each reduction with a call to full(). Assumes we have already
+  // verified that this is safe to do.
+  for (auto rop : ir_utils::getReductionOps(fusion)) {
+    ir_utils::replaceReductionWithFull(rop);
```
Comment (jacobhinkle) [inline csrc/scheduler/pointwise.cpp:440]: @naoyam , as we discussed this replaces reductions with full (after verifying that this is safe during `canScheduleRuntime`). The problem is that this replacement breaks the path from outputs to inputs. In a simple linear kernel such as in the test, `schedulePointwise` fails to find a reference tensor. In this kind of case, we actually have an unused input after the transformation, but I don't think there's a convenient way to eliminate that upstream branch from the replaced expression. Perhaps this could be done by introducing a new OptOutMutator that would remove upstream branches (if this doesn't already exist), since this would clear the uses() for any inputs that would have only dead branches, so the reference tensor would be cleared.

Comment (naoyam) [reply csrc/scheduler/pointwise.cpp:440]: Sounds strange to me that an unused input would the reference detection fail. Unused inputs would happen without this translation, and I think they would be just ignored, no?

Comment (jacobhinkle) [reply csrc/scheduler/pointwise.cpp:440]: Actually, I think I can just use `ir_utils::replaceValue` for this.

Comment (jacobhinkle) [reply csrc/scheduler/pointwise.cpp:440]: It becomes unused after we break the fusion at the reduction. However, the `uses()` of the input is still non-empty because that upstream branch is not cleared. Same happens if I use `ir_utils::replaceValue`.

Comment (jacobhinkle) [reply csrc/scheduler/pointwise.cpp:440]: `OptOutMutator` only does the replacements/removals we register, so we need to do another pass to remove the upstream branch.

File: csrc/ir_utils.cpp [hunk 3]
```diff
@@ -572,6 +574,45 @@ bool isReductionTvOp(const Expr* expr) {
...
+  if (expr->isA<ReductionOp>()) {
+    // Fill value is reduction starting value
+    auto rop = expr->as<ReductionOp>();
+    auto out_root = rop->out()->as<TensorView>()->domain()->root();
+    auto out_shape = std::vector<Val*>(out_root.size());
+    for (auto i : c10::irange(out_root.size())) {
+      out_shape[i] = out_root[i]->extent();
+    }
+    auto new_out = full(out_shape, rop->init(), rop->out()->dtype());
+    ir_utils::replaceValInExpr(expr, rop->out(), new_out);
```
Comment (naoyam) [inline csrc/ir_utils.cpp:605]: We don't want to use `expr` anymore as it's a reduction op.

Instead, try this:

```
auto rop_out = rop->out();
auto out_uses = rop_out->uses();
for (auto out_use: out_uses) {
  ir_utils::replaceValInExpr(out_use, rop_out, new_out);
}
// At this point, rop_out is no longer used by any expr. 
rop->container()->removeVal(rop);
```

I believe this should remove the original reduction expr and its output from the fusion. It should also trigger updating the usage map, so the input should become unused.

Let me know if this works.

Comment (jacobhinkle) [reply csrc/ir_utils.cpp]: Yeah you're right. I pushed a recursive version that removes the branch. After that the test passes. I am now extending this to Welford in which case instead of full we have for example inAvg / inN, which may be nan, matching pytorch.

Comment (naoyam) [reply csrc/ir_utils.cpp]: Just want to understand, is this recursive removal necessary?

Comment (jacobhinkle) [reply csrc/ir_utils.cpp]: I think it's necessary, since there could be ops before the reduction. For example, the test currently reads:
```c++
  auto tv0 = makeSymbolicTensor(2);
  fusion->addInput(tv0);
  auto tv2 = sum(tv0, {1});
  fusion->addOutput(tv2);
```
Removing the sum makes it so that `tv0->uses()` is empty so that it doesn't complain here: https://github.com/NVIDIA/Fuser/blob/main/csrc/scheduler/pointwise_utils.cpp#L109

If we have another op before the reduction:
```c++
  auto tv0 = makeSymbolicTensor(2);
  fusion->addInput(tv0);
  auto tv1 = neg(tv0);
  auto tv2 = sum(tv1, {1});
  fusion->addOutput(tv2);
```
and we just remove the sum, `tv0` would still have a use, so it would block `tv2` being a valid reference tensor.

Comment (naoyam) [reply csrc/ir_utils.cpp]: It shouldn't be. If the sum is removed, the fusion would look like:

```
  auto tv0 = makeSymbolicTensor(2);
  fusion->addInput(tv0);
  auto tv1 = neg(tv0);
  auto tv3 = full(...);
  fusion->addOutput(tv3);
```

Then, the use map should be updated and the neg should no longer appear as a use of `tv0` since it's not connected with the output. I think its allocated objects are still kept in the fusion container, so removing them is helpful, but shouldn't be necessary.

Comment (jacobhinkle) [reply csrc/ir_utils.cpp]: Your right that `resetTvUses()` updates uses for us, as you say. I removed the prune function. I am now trying to fix another issue that came up when I added more ops: finding reference tensor looks at all input IterDomains to see that they map to the reference, but since we accept reductions now those reduced IterDomains in the input are  not mapped to the reference. So I'm updating `DomainMap` to make exceptions for reduced domains in the input.

File: csrc/ir_utils.cpp [hunk 4]
```diff
@@ -572,6 +574,83 @@ bool isReductionTvOp(const Expr* expr) {
...
+}
+
+void replaceReductionWithFull(Expr* expr) {
+  auto fusion = expr->container()->as<Fusion>();
+  if (expr->isA<ReductionOp>()) {
+    // Fill value is reduction starting value
+    auto rop = expr->as<ReductionOp>();
+    auto old_out = rop->out()->as<TensorView>();
+    auto new_out = full_like(old_out, rop->init());
+    ir_utils::replaceValue(fusion, {{old_out, new_out}});
```
Comment (naoyam) [inline csrc/ir_utils.cpp:620]: This may be fine, but would like to confirm if my understanding is correct.

`replaceValue` just does replacements based on the given mappings. I'm not 100% sure, but this seems like it'd apply a replacement to the reduction op, resulting in a new reduction op from the same input to `new_out`. The full op would still be there. All the uses of `old_out` would be replaced with `new_out`, so in the end I think this would result in what we would like, but the replacement of the reduction op should be avoided (if it actually happened) since the new output would not be a valid reduction output tensor.

Comment (naoyam) [reply csrc/ir_utils.cpp:620]: I'd think this would be better:

```
for (auto use: old_out->uses()) {
  ir_utils::replaceValInExpr(use, old_out, new_out);
}
```

File: csrc/scheduler/pointwise.cpp (hunk 2 above)
Comment (naoyam) [reply csrc/scheduler/pointwise.cpp:440]: This is no longer an issue, right?

File: csrc/scheduler/reduction.cpp (hunk 1 above)
Comment (naoyam) [reply csrc/scheduler/reduction.cpp:80]: Does this still happen? Don't the pointwise and noop schedulers take care of all these conditions?

File: csrc/scheduler/registry.cpp [hunk 5]
```diff
@@ -1720,6 +1740,13 @@ class PointWiseScheduler : public SchedulerEntry {
       return !reason.empty();
     }
 
+    // Reject if there is a non-trivial reduction
```
Comment (naoyam) [inline csrc/scheduler/registry.cpp:1743]: Why is this here?

File: csrc/scheduler/registry.cpp [hunk 6]
```diff
@@ -1296,7 +1296,12 @@ class NoOpScheduler : public SchedulerEntry {
         auto all_nonzero = std::none_of(
             concrete_dimension.begin(),
             concrete_dimension.end(),
-            [](IterDomain* id) { return id->extent()->isZeroInt(); });
+            [](IterDomain* id) {
+              // Purely symbolic extents _might_ be zero depending on inputs.
+              // These will need to be checked at runtime. So we should only
+              // reject if the extent is known to be non-zero at compile time.
+              return !id->extent()->isConstInt() || id->extent()->isZeroInt();
```
Comment (naoyam) [inline csrc/scheduler/registry.cpp:1303]: Shouldn't this check be also done at the run time? The compile-time check only makes sure there may be a size-0 non-reduction domain, but it's not guaranteed.

File: csrc/scheduler/registry.cpp [hunk 7]
```diff
@@ -1504,6 +1523,15 @@ class ReductionScheduler : public SchedulerEntry {
       Fusion* fusion,
       SchedulerRuntimeInfo& runtime_info,
       HeuristicSummary* data_cache = nullptr) {
+    for (auto rop : ir_utils::getReductionOps(fusion)) {
+      // Reject if there are any zero-size reductions, since these would prefer
+      // pointwise, or if there are reductions with zero-size concrete
+      // dimensions, since these would prefer NoOp
+      if (ir_utils::isReductionOverSizeZero(rop, runtime_info) ||
```
Comment (naoyam) [inline csrc/scheduler/registry.cpp:1530]: What would happen if there are two reduction domains and only one of them is size-0? It doesn't seem like a full op as there's still a non-size-0 reduction domain, but `isReductionOverSizeZero` would return `true`.

File: csrc/scheduler/registry.cpp [hunk 8]
```diff
@@ -1686,14 +1714,6 @@ class PointWiseScheduler : public SchedulerEntry {
       }
     }
 
-    auto reduction_ops = ir_utils::getReductionOps(fusion);
```
Comment (naoyam) [inline csrc/scheduler/registry.cpp:1689]: Where's the corresponding check of the eligibility of using the pointwise scheduler with reductions?

File: csrc/scheduler/pointwise_utils.cpp [hunk 9]
```diff
@@ -50,6 +50,22 @@ bool DomainMap::areAllInputIdsMappedTo(TensorView* input_tv, TensorView* tv)
   eraseIfInputMappedThroughRFactorDomain(
       in_concrete_ids, tv->getMaybeRFactorDomain());
 
+  // Erase input concrete IDs mapped to any reduction domains
```
Comment (naoyam) [inline csrc/scheduler/pointwise_utils.cpp:53]: I think we can just ignore any reduction domains as they should not show up in the fusion after reduction-to-full translations.

File: csrc/ir_utils.h [hunk 10]
```diff
@@ -15,6 +15,8 @@
 
 namespace nvfuser {
 
+class SchedulerRuntimeInfo;
```
[the remark this answers is not in the document]
Comment (jacobhinkle) [reply csrc/ir_utils.h:18]: I currently only use it for the `ExpressionEvaluator`, so that we can tell which extents are actually zero.

File: csrc/scheduler/registry.cpp (hunk 5 above)
Comment (jacobhinkle) [reply csrc/scheduler/registry.cpp:1743]: These checks get run by `FusionKernelRuntime` when we have previously accepted a Fusion and need to check it against new inputs. If we first run with size-zero reductions but then pass inputs that make those reductions non-trivial, we need Pointwise and NoOp to reject. Same goes for the other direction. All these checks are avoided if we handle reductions at concretization instead since we can then assume the reduction is nontrivial in both concrete and reduction domains.

File: csrc/scheduler/registry.cpp (hunk 7 above)
Comment (jacobhinkle) [reply csrc/scheduler/registry.cpp:1530]: My understanding may be flawed, but in that case there are still zero elements being reduced aren't there? For example we could merge all the reduction IterDomains and would wind up with an extent-zero reduction domain.

File: csrc/scheduler/registry.cpp (hunk 6 above)
Comment (jacobhinkle) [reply csrc/scheduler/registry.cpp:1303]: Yes it needs a runtime check too.

File: csrc/scheduler/registry.cpp (hunk 5 above)
Comment (naoyam) [reply csrc/scheduler/registry.cpp:1743]: Ah, sorry, I thought this was added the transpose scheduler since I saw `TransposeScheduler` at around line 1735.

File: csrc/ir_utils.h [hunk 11]
```diff
@@ -338,6 +340,29 @@ TORCH_CUDA_CU_API bool isReductionOp(const Expr*);
 // Returns if Expr is a reduction op with TensorView or TensorIndex
 TORCH_CUDA_CU_API bool isReductionTvOp(const Expr*);
 
+//! Returns if Expr is a reduction over at least one size-0 dimension
+TORCH_CUDA_CU_API bool isReductionOverSizeZero(
```
[the remark this answers is not in the document]
Comment (csarofeen) [reply csrc/ir_utils.h:344]: Agreed, schedulers commonly use runtime info and we should avoid that leaking out of the schedulers dir.

All of these functions seem to be used around scheduling/segmentation. I don't think it makes sense to have them hang out here.

File: csrc/ir_utils.cpp (hunk 4 above)
Comment (csarofeen) [reply csrc/ir_utils.cpp:620]: I think you're saying the only difference between using `ir_utils::replaceValue` and your proposed loop is that the former would change `rop` to have an incorrect output, and the latter wouldn't touch `rop`.

This is because even though `rop` shouldn't be used since it's disconnected from the outputs, it's still not great to have an illegally defined reduction operation hanging out.

File: csrc/scheduler/pointwise.cpp (hunk 2 above)
Comment (csarofeen) [reply csrc/scheduler/pointwise.cpp:440]: Should be.

File: csrc/scheduler/pointwise_utils.cpp (hunk 9 above)
Comment (csarofeen) [reply csrc/scheduler/pointwise_utils.cpp:53]: Couldn't you simply ignore any input tensor view that doesn't have any `uses()`?

File: csrc/scheduler/reduction.cpp (hunk 1 above)
Comment (csarofeen) [reply csrc/scheduler/reduction.cpp:80]: I have the same question.

File: csrc/scheduler/registry.cpp (hunk 6 above)
Comment (csarofeen) [reply csrc/scheduler/registry.cpp:1303]: I wonder if we should treat zero size as a special type like we do with one size so if a size changes in/out of zero we always recompile. cc @jjsjann123

File: csrc/scheduler/registry.cpp [hunk 12]
```diff
@@ -1335,11 +1329,36 @@ class NoOpScheduler : public SchedulerEntry {
...
+                .as<int64_t>() != 0) {
+          return false;
+        }
+      }
+    }
+
     return true;
   }
 
   void schedule(Fusion* fusion) override {
```
Comment (csarofeen) [inline csrc/scheduler/registry.cpp:1351]: TODO: Delete prints

File: csrc/scheduler/registry.cpp [hunk 13]
```diff
@@ -1335,11 +1329,36 @@ class NoOpScheduler : public SchedulerEntry {
...
-    // Schedule is no-op.
+    // Schedule is no-op, but we need to remove problematic ops. In particular,
+    // reductions with zero concrete size should be converted to full ops.
+    fusion->print();
+    std::cout << "\nRemoving reductions\n" << std::endl;
+    for (auto rop : fusion->exprs()) {
+      if (rop->isA<ReductionOp>() || rop->isA<WelfordOp>()) {
+        ir_utils::replaceReductionWithFull(rop);
```
Comment (csarofeen) [inline csrc/scheduler/registry.cpp:1358]: When would this get hit? Why do we need to remove the reduction ops in a no schedule situation? Is this because of some runtime check?

File: csrc/scheduler/registry.cpp (hunk 7 above)
Comment (csarofeen) [reply csrc/scheduler/registry.cpp:1530]: I agree with this interpretation.

File: csrc/scheduler/registry.cpp (hunk 5 above)
Comment (csarofeen) [reply csrc/scheduler/registry.cpp:1743]: If we made size==0 special like size==1 then we wouldn't need this check.

File: csrc/scheduler/registry.cpp (hunk 8 above)
Comment (csarofeen) [reply csrc/scheduler/registry.cpp:1689]: WDYM?

**Review by csarofeen — COMMENTED** [review]
Could you please summarize the expected behavior of size==0 scenarios you're supporting with this PR?

This feels like a bit strange of a WAR, but I don't see any reason it wouldn't work.

**Comment by jacobhinkle** [comment]
@csarofeen before I address your points point-by-point, note that in some side conversations, @naoyam and I have discussed an alternative to this PR wherein we mark most reduction outputs as `Symbolic`, then perform the replacement with `full` at concretization. That means none of this size-zero reduction logic needs to be present in the schedulers, since we can assume no reductions have size-zero inputs.

Going that route means a large proportion of our `Fusion`s may become dynamic that were otherwise static, but as we see in this current PR static `Fusion`s can also have some complicated rewriting logic during scheduling. Also note that if we can prove the extents are non-zero, that we could potentially avoid dynamic Fusions even with reductions (see #340). So I am leaning toward abandoning this PR and creating a new one for dynamic reductions, once #258 and #244 are merged.

File: csrc/scheduler/pointwise_utils.cpp (hunk 9 above)
Comment (jacobhinkle) [reply csrc/scheduler/pointwise_utils.cpp:53]: This function needs to check that all input IDs are represented by the reference tensor. Any input with empty `uses()` would violate that since its IDs would not map to any others. You're right that this is safe, and is already in place at the one call-site: https://github.com/NVIDIA/Fuser/blob/d0214274982a8d40c2bfeb5047cfc5bd130379ad/csrc/scheduler/pointwise_utils.cpp#L124-L127. Note that more generally, we could use `InputsOf(fusion->outputs())` so that empty uses propagate back to give us more ignored inputs.

The check here is actually skipping any input IDs that map to internal reductions (not represented in outputs. That seemed like a good idea at the time but now I realize that's a bad idea, since we do still need those IDs represented in the reference tensor.

File: csrc/ir_utils.cpp [hunk 14]
```diff
@@ -572,6 +574,83 @@ bool isReductionTvOp(const Expr* expr) {
...
+      if (!id->isReduction() &&
+          expr_eval.evaluate(id->extent())->as<int64_t>() == 0) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+void replaceReductionWithFull(Expr* expr) {
```
[the remark this answers is not in the document]
Comment (jacobhinkle) [reply csrc/ir_utils.cpp:613]: Actually `prod` is covered the same as `sum`: by replacing the op with `full(init_val)`. PyTorch is consistent with this:
```
[ins] In [2]: torch.prod(torch.randn([3,0]))
Out[2]: tensor(1.)
```

**Comment by jacobhinkle** [comment]
> Could you please summarize the expected behavior of size==0 scenarios you're supporting with this PR?

I mentioned it above in some of the responses, but basically I think `size == 0` has two key implications:
- TVs with size-0 domains do not allocations and do not need to compute their definitions
- Reductions over size-0 domains are equivalent to a constant (the initial value of the reduction fold).

> This feels like a bit strange of a WAR, but I don't see any reason it wouldn't work.

It definitely doesn't feel clean. Maybe it's because it's the shiny new toy, but I'm inclined to use the dynamic fusion machinery for this instead.

File: csrc/ir_utils.cpp [hunk 15]
```diff
@@ -572,6 +574,83 @@ bool isReductionTvOp(const Expr* expr) {
...
+    auto rop = expr->as<ReductionOp>();
+    auto old_out = rop->out()->as<TensorView>();
+    auto new_out = full_like(old_out, rop->init());
+    ir_utils::replaceValue(fusion, {{old_out, new_out}});
+    if (old_out->isFusionOutput()) {
+      fusion->replaceOutput(old_out, new_out);
+    }
+  } else if (expr->isA<WelfordOp>()) {
+    // Fill Welford avg/var with nan, N with 0
+    auto wop = expr->as<WelfordOp>();
```
[the remark this answers is not in the document]
Comment (jacobhinkle) [reply csrc/ir_utils.cpp:626]: Yes I think it could. I originally had a little more logic in here that was wrong, so it was removed. Each output of a Welford is equal to the two reductions each divided by `N`. For size-0 reduction domain, this is 0/0. So we just need to add the division by 0 in the case of Welford, but otherwise it's the same logic as for any other reduction.

**Comment by jjsjann123** [comment]
> @naoyam and I have discussed an alternative to this PR wherein we mark most reduction outputs as Symbolic, then perform the replacement with full at concretization. That means none of this size-zero reduction logic needs to be present in the schedulers, since we can assume no reductions have size-zero inputs.

This sounds like a much better approach to me than trying to work around this corner case in scheduler.

> I wonder if we should treat zero size as a special type like we do with one size so if a size changes in/out of zero we always recompile. cc @jjsjann123

I think the specialization here can also be worked into concretization, and is likely better that way instead of polluting our existing API. I'm expecting size-0 to not change frequently across runs, so hopefully this won't trigger extended overhead and recompilation.
In the long run, the holy grail is to have concretization handling both shape specialization as well as contiguity/stride_order specialization, so we can hide those runtime info from user API when defining computation.

**Comment by jacobhinkle** [comment]
Closing in favor of #449
